src-os/ptime_clock_stubs.c

```diff
@@ -106,6 +110,45 @@ CAMLprim value ocaml_ptime_clock_now_d_ps (value unit)
   CAMLreturn (pair);
 }
 
+#elif defined(OCAML_PTIME_WIN)
+
+CAMLprim value ocaml_ptime_clock_now_d_ps (value unit)
+{
+  CAMLparam1 (unit);
+  CAMLlocal1 (pair);
+  long sec, usec;
+  SYSTEMTIME stime;
+  FILETIME ftime;
+  uint64_t time;
+
+  static const uint64_t epoch = ((uint64_t)116444736000000000ULL);
+
+  GetSystemTime (&stime);
+  SystemTimeToFileTime (&stime, &ftime);
+  time = ((uint64_t)ftime.dwLowDateTime) +
+         (((uint64_t)ftime.dwHighDateTime) << 32);
+
+  sec = (long)((time - epoch) / 10000000L);
+  usec = (long)(stime.wMilliseconds * 1000);
+
+  if (usec < 0 || usec > 999999)
+    OCAML_PTIME_RAISE_SYS_ERROR ("unreasonable usec in timeval");
+
+  if (sec < 0)
+    OCAML_PTIME_RAISE_SYS_ERROR ("negative sec in timeval");
+
+  int d = sec / 86400;
+  if (d > OCAML_PTIME_DAY_MAX)
+    OCAML_PTIME_RAISE_SYS_ERROR ("can't represent timeval in Ptime.t");
```

Comment:
These tests now appear in three places (and, for what it's worth, the message is slightly different in one). Might it be worth factoring this out to a macro so you can say something like `OCAML_PTIME_VALIDATE(usec, sec);`?

Reply:
Right. @IndiscriminateCoding no need to do this I'll have a look post merge.